staging/src/k8s.io/component-helpers/storage/volume/helpers_test.go

```diff
@@ -206,6 +207,112 @@ func TestCheckVolumeNodeAffinity(t *testing.T) {
 	}
 }
 
+func testVolumeWithCSINodeDriverAffinity(t *testing.T, volumeSource *v1.CSIPersistentVolumeSource) *v1.PersistentVolume {
+	return &v1.PersistentVolume{
+		ObjectMeta: metav1.ObjectMeta{Name: "test-constraints"},
+		Spec: v1.PersistentVolumeSpec{
+			PersistentVolumeSource: v1.PersistentVolumeSource{
+				CSI: volumeSource,
+			},
+		},
+	}
+}
+
+func testCSINode(t *testing.T, csiDrivers []storageV1.CSINodeDriver) *storageV1.CSINode {
+	return &storageV1.CSINode{
+		ObjectMeta: metav1.ObjectMeta{Name: "test-csinodes"},
+		Spec: storageV1.CSINodeSpec{
+			Drivers: csiDrivers,
+		},
+	}
+}
+
+func TestCheckVolumeCSINodeDriverAffinity(t *testing.T) {
+	type csiNodeDriverAffinityTest struct {
+		name          string
+		expectSuccess bool
+		pv            *v1.PersistentVolume
+		csiNode       *storageV1.CSINode
+	}
+
+	cases := []csiNodeDriverAffinityTest{
+		{
+			name:          "valid-pv-nil",
+			expectSuccess: true,
+			pv:            testVolumeWithCSINodeDriverAffinity(t, nil),
+			csiNode: testCSINode(t, []storageV1.CSINodeDriver{
+				{
+					Name: "test-driver",
+				},
+				{
+					Name: "test-driver2",
+				},
+			}),
+		},
+		{
+			name:          "valid-csi-node-nil",
+			expectSuccess: false,
+			pv: testVolumeWithCSINodeDriverAffinity(t, &v1.CSIPersistentVolumeSource{
+				Driver:       "test-driver",
+				VolumeHandle: "diskId",
+			}),
+			csiNode: nil,
+		},
+		{
+			name:          "valid-pv-csinode-mismatch",
+			expectSuccess: false,
+			pv: testVolumeWithCSINodeDriverAffinity(t, &v1.CSIPersistentVolumeSource{
+				Driver:       "test-driver",
+				VolumeHandle: "diskId",
+			}),
+			csiNode: testCSINode(t, []storageV1.CSINodeDriver{
+				{
+					Name: "test-driver2",
+				},
+				{
+					Name: "test-driver3",
+				},
+			}),
+		},
+		{
+			name:          "valid-pv-csinode-match",
+			expectSuccess: true,
+			pv: testVolumeWithCSINodeDriverAffinity(t, &v1.CSIPersistentVolumeSource{
+				Driver:       "test-driver",
+				VolumeHandle: "diskId",
+			}),
+			csiNode: testCSINode(t, []storageV1.CSINodeDriver{
+				{
+					Name: "test-driver",
+				},
+				{
+					Name: "test-driver1",
+				},
+			}),
+		},
+		{
+			name:          "valid-pv-csinode-empty",
+			expectSuccess: false,
+			pv: testVolumeWithCSINodeDriverAffinity(t, &v1.CSIPersistentVolumeSource{
+				Driver:       "test-driver",
+				VolumeHandle: "diskId",
+			}),
+			csiNode: testCSINode(t, []storageV1.CSINodeDriver{}),
+		},
+	}
+
+	for _, c := range cases {
+		err := CheckCSINodeDriverAffinity(c.csiNode, c.pv)
```

Comment:
Do we use expectedErr to make sure the error output is what we expected?

Reply:
Because we have different errors, we need to make sure the errors are as expected.

Reply:
Sounds good, I'll add it in next commit, I thought at first it might be too simple to need separate error handling,